### PR TITLE
[WIP] Simplify dagger modules

### PR DESCRIPTION
**Pull request by wsargent**
Uses the built in components to leverage Play application.

Still isn't all the way through as the body parser is failing on injection.

UPDATE: Got the body parser to work, however there is still some global state being accessed:

```
[ERROR] [04/28/2017 21:06:51.458] [play-dev-mode-akka.actor.default-dispatcher-4] [akka.actor.ActorSystemImpl(play-dev-mode)] Internal server error, sending 500 response
java.lang.RuntimeException: The global application is disabled. Set play.allowGlobalApplication to allow global state here.
	at scala.sys.package$.error(package.scala:27)
	at play.api.Play$.privateMaybeApplication(Play.scala:86)
	at play.api.http.HttpConfiguration$.current(HttpConfiguration.scala:259)
	at play.api.mvc.Cookies$.config(Cookie.scala:139)
	at play.api.mvc.CookieHeaderEncoding.encodeSetCookieHeader(Cookie.scala:210)
	at play.api.mvc.CookieHeaderEncoding.encodeSetCookieHeader$(Cookie.scala:209)
	at play.api.mvc.Cookies$.encodeSetCookieHeader(Cookie.scala:136)
	at play.api.mvc.Result.bakeCookies(Results.scala:295)
	at play.core.server.common.ServerResultUtils.prepareCookies(ServerResultUtils.scala:229)
	at play.core.server.AkkaHttpServer.$anonfun$executeAction$3(AkkaHttpServer.scala:254)
	at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)
	at akka.http.scaladsl.util.FastFuture$.transformWith$extension1(FastFuture.scala:55)
	at akka.http.scaladsl.util.FastFuture$.flatMap$extension(FastFuture.scala:26)
	at play.core.server.AkkaHttpServer.executeAction(AkkaHttpServer.scala:253)
	at play.core.server.AkkaHttpServer.executeHandler(AkkaHttpServer.scala:208)
	at play.core.server.AkkaHttpServer.handleRequest(AkkaHttpServer.scala:155)
	at play.core.server.AkkaHttpServer.$anonfun$createServerBinding$4(AkkaHttpServer.scala:90)
```

### CODE REVIEW

File: app/dagger/ClockModule.java [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
+/**
+ * A module that provides a clock implementation.
+ */
 @Module
 public class ClockModule {
 
-    @Singleton
     @Provides
-    public Clock providesClock() {
+    public Clock clock() {
```
Comment (marcospereira) [inline app/dagger/ClockModule.java:12]: You won't need a clock module:

https://github.com/playframework/playframework/blob/3bd429d9520c297a037e41edaf7f5a6f77701852/framework/src/play/src/main/java/play/components/CryptoComponents.java#L17-L20

There is a `TODO`, but you can reuse that clock anyway.

File: app/dagger/MyApplicationLoader.java [hunk 2]
```diff
@@ -2,43 +2,20 @@
...
-
-        public void configure(Environment env) {
-            delegate.configure(env.underlying());
-        }
-    }
-
     @Override
-    public Application load(Context context) {
-        final ClassLoader classLoader = context.environment().classLoader();
-        final Optional<LoggerConfigurator> opt = LoggerConfigurator.fromClassLoader(classLoader);
```
Comment (marcospereira) [inline app/dagger/MyApplicationLoader.java]: You still need to call the `LoggerConfigurator` here. See this example:

https://github.com/playframework/playframework/blob/3bd429d9520c297a037e41edaf7f5a6f77701852/documentation/manual/working/javaGuide/main/dependencyinjection/code/javaguide/di/components/CompileTimeDependencyInjection.java#L53-L64

File: app/dagger/MyComponentsFromContext.java [hunk 3]
```diff
@@ -1,109 +1,59 @@
...
-    @Override
-    public play.api.inject.Injector injector() {
-        // We need to add any Java actions and body parsers needed to the runtime injector
-        return new SimpleInjector(super.injector(), Scala.asScala(new HashMap<Class<?>, Object>() {{
-            put(play.mvc.BodyParser.Default.class, new play.mvc.BodyParser.Default(javaErrorHandler(), httpConfiguration(), playBodyParsers()));
-            put(play.api.i18n.MessagesApi.class, messagesApi);
-            put(play.api.i18n.Langs.class, langs);
-        }}));
-    }
+        extraMappings.put(JavaHandlerComponents.class, () -> new DefaultJavaHandlerComponents(simpleInjector.asScala(), actionCreator(), httpConfiguration(), executionContext(), javaContextComponents()));
```
Comment (marcospereira) [inline app/dagger/MyComponentsFromContext.java:45]: This probably should be solved by `BuiltInComponentsFromContext` itself.

File: app/dagger/SimpleInjector.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import scala.reflect.ClassTag;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * A simple injector with additional classes...
+ */
+public class SimpleInjector implements Injector {
```
Comment (marcospereira) [inline app/dagger/SimpleInjector.java:13]: Can't we use the Scala [`SimpleInjector`](https://github.com/playframework/playframework/blob/8d543c6a476a306f556772d8682b867104028090/framework/src/play/src/main/scala/play/api/inject/Injector.scala#L64-L104) instead?

File: app/dagger/ClockModule.java (hunk 1 above)
Comment (wsargent) [reply app/dagger/ClockModule.java:12]: This is to demo the Clock injection though

File: app/dagger/SimpleInjector.java (hunk 4 above)
Comment (wsargent) [reply app/dagger/SimpleInjector.java:13]: I'm not sure if that's any better -- we have a java injector() method that would call out to a scala one, and then convert it back to java...

**Pull request merged**